nautobot/docs/user-guide/administration/upgrading/from-v1/ipam-upgrade-guide.md

```diff
@@ -0,0 +1,191 @@
+- If the `Prefix.type` is `Network`:
+    - The utilization is calculated as the sum of the total address space of all child `Pool` prefixes plus the total number of child IP addresses.
+    - For IPv4 networks with a `prefix_length` larger (lower) than `/31`, if neither the first or last address is occupied by either a pool or an IP address, they are subtracted from the total size of the prefix.
+
+#### Example
+
+- 192.168.0.0/16          `Container - 1024/65536 utilization`
+    - 192.168.1.0/24      `Network - 1/254 utilization`
+        - 192.168.1.1     `IP Address`
+    - 192.168.2.0/24      `Network - 4/256 utilization`
+        - 192.168.2.0/30  `Pool - 1/4 utilization`
+            - 192.168.2.1 `IP Address`
+    - 192.168.3.0/24      `Network - 5/254 utilization`
+        - 192.168.3.1     `IP Address`
+        - 192.168.3.64/30 `Pool - 0/4 utilization`
+    - 192.168.4.0/24      `Network - 1/256 utilization`
+        - 192.168.4.255   `IP Address`
+
+### IPAddress Parenting Concrete Relationship was added
+
+The `ipam.IPAddress` model has been modified to have a mandatory foreign key to `ipam.Prefix` as the `parent` field. Parenting of IP addresses is now automatically managed at the database level to greatly improve performance especially when calculating tree hierarchy and utilization.
+
+| Removed                | Replaced With   |
+| ---------------------- | --------------- |
+| `get_child_prefixes()` | `descendants()` |
+
+#### IPAddress Parenting Guidance
+
+The following guidance has been added to the `IPAddress` model:
+
+- An `IPAddress` should have a parent `Prefix` of type `Network`
+- An `IPAddress` should not be created if a suitable parent `Prefix` of type `Network` does not exist
+- An `IPAddress` can be a member of a `Pool` but only if the `Pool` is a child of a `Network`
+- If an eligible parent `Prefix` cannot be found for an `IPAddress` in a given `Namespace`, creation or update of that `IPAddress` will fail.
+- If deleting a parent `Prefix` would result in any child `IPAddress` objects to become orphaned, the delete opration will fail.
+
+As with the [`Prefix` parenting guidance](#prefix-parenting-guidance) above, violating this guidance in Nautobot 2.0 will result in a warning; in a future Nautobot release this will be changed to an enforced data constraint.
+
+### Parenting affinity during the upgrade
+
+A best effort is made to keep `Prefixes` and `IPAddresses` together by shared attributes such as Tenant, but this is not alway possible for various reasons such as numerous duplicate with identical or too-closely-similar criteria.
+
+When identifying possible ancestors for child `Prefix` or `IPAddress` objects during the reparenting phase of the upgrade process, the following sets of attributes will be compared in order:
+
+- `Tenant` assigned, `VRF` assigned
+- `Tenant` null, `VRF` assigned
+- `Tenant` assigned, `VRF` null
+- `VRF` assigned
+- `Tenant` null, `VRF` null
+- `VRF` null
+- Default ordering
```

Comment:
I don't have specific suggestions at the moment for this section but it "feels" not directly useful to the user to me. Maybe some concrete examples might help?

Reply:
Here's the relevant code for reference:

```py
                models.When(tenant=obj.tenant, vrf=obj.vrf, then=models.Value(1)), 
                models.When(tenant__isnull=True, vrf=obj.vrf, then=models.Value(2)), 
                models.When(tenant=obj.tenant, vrf__isnull=True, then=models.Value(3)),
                models.When(vrf=obj.vrf, then=models.Value(4)),
                models.When(tenant__isnull=True, vrf__isnull=True, then=models.Value(5)),
                models.When(vrf__isnull=True, then=models.Value(6)),
                default=models.Value(7),
```

Maybe something like this?

```mermaid
flowchart TD
    8[Prefix 10.0.0.0/8]
    8 --> 24BlueA[Prefix 10.0.0.0/24<br>VRF <font color=00BBFF>Blue</font><br>Tenant A] --> IPBlueA[IPAddress 10.1.1.1<br>VRF <font color=00BBFF>Blue</font><br>Tenant A<br><br><i>Prefer matching tenant and VRF</i>]
    8 --> 24BlueNone[Prefix 10.0.0.0/24<br>VRF <font color=00BBFF>Blue</font><br>Tenant None]  --> IPBlueB[IPAddress 10.1.1.1<br>VRF <font color=00BBFF>Blue</font><br>Tenant B<br><br><i>Prefer matching VRF with null tenant<br>before different tenant</i>]
    8 --> 24NoneB[Prefix 10.0.0.0/24<br>VRF None<br>Tenant B] --> IPGreenB[IPAddress 10.1.1.1<br>VRF <font color=66FF66>Green</font><br>Tenant B<br><br><i>Prefer matching tenant with null VRF<br>before different VRF</i>]
    8 --> 24RedB[Prefix 10.0.0.0/24<br>VRF <font color=red>Red</font><br>Tenant B]  --> IPRedA[IPAddress 10.1.1.1<br>VRF <font color=red>Red</font><br>Tenant A<br><br><i>Prefer matching VRF if no suitable match<br>can be found with matching or null tenant</i>]
    8 --> 24NoneNone[Prefix 10.0.0.0/24<br>VRF None<br>Tenant None] --> IPYellowY[IPAddress 10.1.1.1<br>VRF <font color=FFFF00>Yellow</font><br>Tenant Y<br><br><i>Prefer null VRF if no suitable match<br>can be found with either<br>matching tenant,<br>matching VRF,<br>or both null VRF and tenant</i>]
```